# Развитие индустрии игр в начале XXI века.

1 Загрузка и знакомство с данными  
2 Проверка ошибок в данных и их предобработка  
2.1 Названия, или метки, столбцов датафрейма  
2.2 Типы данных  
2.3 Наличие пропусков в данных  
2.4 Явные и неявные дубликаты в данных  
3 Фильтрация данных  
4 Категоризация данных  
5 Вывод 

## 1 Загрузка и знакомство с данными




In [75]:
##Загрузите необходимые библиотеки Python и данные датасета /datasets/new_games.csv.
import pandas as pd
import numpy as np
df=pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [76]:
#Знакомимся с данными: выводим первые строки и результат метода info().
display(df.head())
display(df.info())

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


None

**Вывод о полученных данных:** всего 16956 строк и 11 столбцов, 
есть пропуски в столбцах Name, Year of Release, Genre, Critic Score, User Score, Rating. 
Года некорректно прописаны надо исправить.

# 2 Проверка ошибок в данных и их предобработка

## 2.1 Названия, или метки, столбцов датафрейма:

In [80]:
#Выведим на экран названия всех столбцов датафрейма и проверьте их стиль написания.
display(df.columns.values)

array(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales',
       'EU sales', 'JP sales', 'Other sales', 'Critic Score',
       'User Score', 'Rating'], dtype=object)

In [81]:
#Приводим все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания _.
def snake_case(x):
    SC=[]
    for index in x.columns.values.tolist():
        SC.append(str(index).lower().replace(' ', '_'))
    return SC
df=df.set_axis(snake_case(df), axis=1)

## 2.2 Типы данных


In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [84]:
df['year_of_release'].unique()

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

In [85]:
df['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', nan, 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION',
       'SHOOTER', 'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE',
       'SIMULATION', 'PUZZLE', 'STRATEGY'], dtype=object)

In [86]:
df['critic_score'].unique()

array([76., nan, 82., 80., 89., 58., 87., 91., 61., 97., 95., 77., 88.,
       83., 94., 93., 85., 86., 98., 96., 90., 84., 73., 74., 78., 92.,
       71., 72., 68., 62., 49., 67., 81., 66., 56., 79., 70., 59., 64.,
       75., 60., 63., 69., 50., 25., 42., 44., 55., 48., 57., 29., 47.,
       65., 54., 20., 53., 37., 38., 33., 52., 30., 32., 43., 45., 51.,
       40., 46., 39., 34., 35., 41., 36., 28., 31., 27., 26., 19., 23.,
       24., 21., 17., 22., 13.])

In [87]:
df['user_score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

In [88]:
df['rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

## 2.3 Наличие пропусков в данных

In [90]:
#Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.
print(f'Количество name NaN значений {df["name"].isna().sum()} и их доля от общего числа {round(df["name"].isna().sum()/len(df["name"]),3)}')
print(f'Количество platform NaN значений {df["platform"].isna().sum()} и их доля от общего числа {round(df["platform"].isna().sum()/len(df["platform"]),3)}')
print(f'Количество year_of_release NaN значений {df["year_of_release"].isna().sum()} и их доля от общего числа {round(df["year_of_release"].isna().sum()/len(df["year_of_release"]),3)}')
print(f'Количество genre NaN значений {df["genre"].isna().sum()} и их доля от общего числа {round(df["genre"].isna().sum()/len(df["genre"]),3)}')
print(f'Количество na_sales NaN значений {df["na_sales"].isna().sum()} и их доля от общего числа {round(df["na_sales"].isna().sum()/len(df["na_sales"]),3)}')
print(f'Количество eu_sales NaN значений {df["eu_sales"].isna().sum()} и их доля от общего числа {round(df["eu_sales"].isna().sum()/len(df["eu_sales"]),3)}')
print(f'Количество jp_sales NaN значений {df["jp_sales"].isna().sum()} и их доля от общего числа {round(df["jp_sales"].isna().sum()/len(df["jp_sales"]),3)}')
print(f'Количество other_sales NaN значений {df["other_sales"].isna().sum()} и их доля от общего числа {round(df["other_sales"].isna().sum()/len(df["other_sales"]),3)}')
print(f'Количество critic_score NaN значений {df["critic_score"].isna().sum()} и их доля от общего числа {round(df["critic_score"].isna().sum()/len(df["critic_score"]),3)}')
print(f'Количество user_score NaN значений {df["user_score"].isna().sum()} и их доля от общего числа {round(df["user_score"].isna().sum()/len(df["user_score"]),3)}')
print(f'Количество rating NaN значений {df["rating"].isna().sum()} и их доля от общего числа {round(df["rating"].isna().sum()/len(df["rating"]),3)}')

Количество name NaN значений 2 и их доля от общего числа 0.0
Количество platform NaN значений 0 и их доля от общего числа 0.0
Количество year_of_release NaN значений 275 и их доля от общего числа 0.016
Количество genre NaN значений 2 и их доля от общего числа 0.0
Количество na_sales NaN значений 0 и их доля от общего числа 0.0
Количество eu_sales NaN значений 0 и их доля от общего числа 0.0
Количество jp_sales NaN значений 0 и их доля от общего числа 0.0
Количество other_sales NaN значений 0 и их доля от общего числа 0.0
Количество critic_score NaN значений 8714 и их доля от общего числа 0.514
Количество user_score NaN значений 6804 и их доля от общего числа 0.401
Количество rating NaN значений 6871 и их доля от общего числа 0.405


In [91]:
df=df.dropna(subset=['year_of_release'])

In [92]:
#Проведение преобразования типов данных.
df['year_of_release']=df['year_of_release'].astype('int64')
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16681 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16681 non-null  object 
 2   year_of_release  16681 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16681 non-null  float64
 5   eu_sales         16681 non-null  object 
 6   jp_sales         16681 non-null  object 
 7   other_sales      16681 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       9974 non-null   object 
 10  rating           9901 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 1.5+ MB


In [93]:
df=df.dropna(subset=['name'])
df=df.dropna(subset=['genre'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16679 non-null  object 
 6   jp_sales         16679 non-null  object 
 7   other_sales      16679 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       9974 non-null   object 
 10  rating           9901 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 1.5+ MB


In [94]:
#Замена значения unknown и преобразования к float 
df['jp_sales']=df['jp_sales'].str.replace('unknown', '-1')
df['jp_sales']= pd.to_numeric(df['jp_sales'])
df['eu_sales']=df['eu_sales'].str.replace('unknown', '-1')
df['eu_sales']= pd.to_numeric(df['eu_sales'])
df.info()




<class 'pandas.core.frame.DataFrame'>
Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16679 non-null  float64
 6   jp_sales         16679 non-null  float64
 7   other_sales      16679 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       9974 non-null   object 
 10  rating           9901 non-null   object 
dtypes: float64(5), int64(1), object(5)
memory usage: 1.5+ MB


In [95]:
df['user_score'] = df['user_score'].replace('tbd', np.nan)
df['user_score']= pd.to_numeric(df['user_score'])
#df["user_score"] = df.groupby(["platform","year_of_release"])["user_score"].transform(lambda x: x.fillna(x.mean()))

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16679 non-null  float64
 6   jp_sales         16679 non-null  float64
 7   other_sales      16679 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       7558 non-null   float64
 10  rating           9901 non-null   object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.5+ MB


**Вывод:** Все столбцы кроме critic_score, user_score, rating без пропусков, эти столбцы мы заполнить не можем потому что они заполняются подьзователями, если необходимо заполнить можно заполнить 0/-1, удалить не можем потому что составляет до половины выборки.  

## 2.4 Явные и неявные дубликаты в данных:

In [99]:
#Подсчет количества пропусков 
display(f'Количество явных дубликатов {df.duplicated(keep="last").sum()} и относительное количество явных дубликатов {round(df.duplicated(keep="last").sum()/len(df),3)}')

'Количество явных дубликатов 179 и относительное количество явных дубликатов 0.011'

In [100]:
for i in ['genre', 'platform', 'rating', 'year_of_release']:
    display(df[i].value_counts())

genre
Action          3341
Sports          2325
Misc            1740
Role-Playing    1493
Shooter         1314
Adventure       1309
Racing          1244
Platform         891
Simulation       866
Fighting         844
Strategy         679
Puzzle           577
ACTION            12
SPORTS             7
ROLE-PLAYING       6
RACING             6
FIGHTING           6
SHOOTER            5
ADVENTURE          4
MISC               3
PLATFORM           3
PUZZLE             2
SIMULATION         1
STRATEGY           1
Name: count, dtype: int64

platform
PS2     2154
DS      2147
PS3     1330
Wii     1305
X360    1249
PSP     1212
PS      1208
PC       972
GBA      826
XB       818
GC       549
3DS      522
PSV      434
PS4      395
N64      320
XOne     251
SNES     241
SAT      174
WiiU     147
2600     118
NES      100
GB        97
DC        52
GEN       27
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: count, dtype: int64

rating
E       3968
T       2948
M       1558
E10+    1414
EC         8
K-A        3
AO         1
RP         1
Name: count, dtype: int64

year_of_release
2009    1450
2008    1445
2010    1279
2007    1218
2011    1149
2006    1020
2005     950
2002     839
2003     789
2004     771
2012     670
2015     612
2014     584
2013     552
2016     507
2001     491
1998     384
2000     357
1999     341
1997     293
1996     267
1995     220
1994     121
1993      60
1981      46
1992      43
1991      42
1982      37
1986      22
1983      18
1987      17
1989      17
1990      16
1988      15
1985      14
1984      14
1980       9
Name: count, dtype: int64

In [101]:
df['name']=df['name'].str.lower()

In [102]:
df['genre']=df['genre'].str.lower()

In [103]:
df['rating']=df['rating'].str.upper()

In [104]:
#Удаление явных дубликатов
df.drop_duplicates(inplace=True)

**Вывод:** По данным выявили, что есть по жанрам различающиеся в написании неявные дубликаты. Пример:Action ACTION, Shooter SHOOTER. Также некотрые платформы очень похожи в описании PS2 PS4. 


</div>
<div class="alert alert-success">

**Артём:** ✔️ Хорошо. Хотя стоило сделать проверку было/стало.

# 3 Фильтрация данных

In [108]:
#Отобраны данные по показателю период с 2000 по 2013
df_actual=df[(df['year_of_release']>=2000)&(df['year_of_release']<=2013)].copy()

# 4 Категоризация данных

In [110]:
#Разделите все игры по оценкам пользователей и выделите такие категории: 
#высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и 
#[](от 0 до 3, не включая правую границу интервала).
def user_sc(x):
    if x['user_score']>=8 and x['user_score']<=10:
        return 'высокая оценка'
    elif x['user_score']>=3 and x['user_score']<8:
        return 'средняя оценка'
    elif x['user_score']>=0 and x['user_score']<3: 
        return 'низкая оценка'
    else:
        return 'без оценки'
df_actual['games_user_categorization']=df_actual.apply(user_sc, axis=1)

In [111]:
#Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), 
#средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).
def critic_sc(x):
    if x['critic_score']>=80 and x['critic_score']<=100:
        return 'высокая оценка'
    elif x['critic_score']>=30 and x['critic_score']<80:
        return 'средняя оценка'
    elif x['critic_score']>=0 and x['critic_score']<30: 
        return 'низкая оценка' 
    else:
        return 'без оценки'
df_actual['games_critic_categorization']=df_actual.apply(critic_sc, axis=1)

In [112]:
df_actual.head(10)

name platform  year_of_release       genre  na_sales  \
0                  wii sports      Wii             2006      sports     41.36   
2              mario kart wii      Wii             2008      racing     15.68   
3           wii sports resort      Wii             2009      sports     15.61   
6       new super mario bros.       DS             2006    platform     11.28   
7                    wii play      Wii             2006        misc     13.96   
8   new super mario bros. wii      Wii             2009    platform     14.44   
10                 nintendogs       DS             2005  simulation      9.05   
11              mario kart ds       DS             2005      racing      9.71   
13                    wii fit      Wii             2007      sports      8.92   
14         kinect adventures!     X360             2010        misc     15.00   

    eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0      28.96      3.77         8.45          76.0         8.0      E   
2      12.76      3.79         3.29          82.0         8.3      E   
3      10.93      3.28         2.95          80.0         8.0      E   
6       9.14      6.50         2.88          89.0         8.5      E   
7       9.18      2.93         2.84          58.0         6.6      E   
8       6.94      4.70         2.24          87.0         8.4      E   
10     10.95      1.93         2.74           NaN         NaN    NaN   
11      7.47      4.13         1.90          91.0         8.6      E   
13      8.03      3.60         2.15          80.0         7.7      E   
14      4.89      0.24         1.69          61.0         6.3      E   

   games_user_categorization games_critic_categorization  
0             высокая оценка              средняя оценка  
2             высокая оценка              высокая оценка  
3             высокая оценка              высокая оценка  
6             высокая оценка              высокая оценка  
7             средняя оценка              средняя оценка  
8             высокая оценка              высокая оценка  
10                без оценки                  без оценки  
11            высокая оценка              высокая оценка  
13            средняя оценка              высокая оценка  
14            средняя оценка              средняя оценка

In [113]:
#Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период.
df_actual.groupby('platform').agg({'name':'count'}).apply(lambda x: x.sort_values(ascending=False).head(7))

name
platform      
PS2       2127
DS        2120
Wii       1275
PSP       1180
X360      1121
PS3       1087
GBA        811

**Вывод:** На финальном этапе анализировались данные за 2000-2013 годы, были удалены записи с пропусками в столбцах name,
genre, year_of_release в которых были пропуски. Заменены unknown на 0, в столбцах jp_sales, eu_sales. Были удалены явные дубликаты 
в количеситве 179 записей.
Добавлены новые поля games_user_categorization, games_critic_categorization, подсчитана сводная статистика по наиболее популярным платформам. 